### ISSUE TITLE
Implementation of the video demo in Unity

### ISSUE THREAD

**Issue by michaelneri** [eth-siplab/AvatarPoser#30]
First of all, thank you for your work.

I am currently working on implementing the demo video you provided in Unity using a VR headset (Meta Quest Pro). Specifically, I want to use your model in real-time to infer the human pose from the HMD and the controllers.

 The data collection procedure is performed at each frame in Unity as follows:

```
        StartCoroutine(DelayAction(0.05f));
        Vector3[] currentPositions = new Vector3[3]; // 0: HMD, 1: Left Controller, 2: Right Controller
        Matrix4x4[] currentRotations = new Matrix4x4[3]; // 0: HMD, 1: Left Controller, 2: Right Controller
        // Collect HMD data
        if (HMD != null)
        {
            currentPositions[0] = ConvertTranslationFromUnityToMaya(HMD.transform.position);
            Quaternion temp = ToRightHanded(HMD.transform.localRotation);
            (float angle, Vector3 axis) = fromQuatToAngleAxis(temp);
            currentRotations[0] = fromAngleAxisToSO(angle, axis);
        }
        // Collect left hand data
        if (leftController != null)
        {
            currentPositions[1] = ConvertTranslationFromUnityToMaya(leftController.transform.position);
            Quaternion temp = ToRightHanded(leftController.transform.localRotation);
            (float angle, Vector3 axis) = fromQuatToAngleAxis(temp);
            currentRotations[1] = fromAngleAxisToSO(angle, axis);
        }
        // Collect right hand data
        if (rightController != null)
        {
            currentPositions[2] = ConvertTranslationFromUnityToMaya(rightController.transform.position);
            Quaternion temp  = ToRightHanded(rightController.transform.localRotation);
            (float angle, Vector3 axis) = fromQuatToAngleAxis(temp);
            currentRotations[2] = fromAngleAxisToSO(angle, axis);
        }

```

where I collect both positions and rotations of HMD and controllers at 60Hz, performing both conversion of coordinates and extracting the rotation matrix associated to it. However, the final result is an avatar that performs badly basic movements. I am using the ONNX version of your model with pretrained weights.

How was your inference code using the headset? Did you have the code to reproduce it?

Thanks in advance for the response!

**Comment by michaelneri** [comment]
I solved it, I was using angle-axis angles rather than Rodrigues ones. Using them, the model works as expected!

**Comment by wwwpkol** [comment]
Dear michaelneri,

Currently, I am using HTC Vive Tracking to capture positions and rotations. However, directly feeding them into the network does not yield good results; I suspect it might be a calibration issue. I noticed that your research is closely related to my work. Could you please give me some advice or share calibration code with me? Below are parts of my data and preprocessing code.
`if __name__ == '__main__':

    # h2r  0:head 1:right_hand 2:left_hand  3:tracker_1 Root  tracker_2:left_foot  tracker_3:right_foot   #linger
    dst = 'D:\code\data\VR_6F_H2R\process'
    for npz_file in npz_files:
        file_path = os.path.join(cfg.data_path, npz_file)
        data_all = dict()
        data = np.load(file_path)['pose']
        # data_all.append(data['pose'])
        seq = data.shape[0]
        feature_bz = 6
        #head
        hmd_trans = torch.from_numpy((data[:, :3]))
        hmd_rot = torch.from_numpy((np.radians(data[:, 3:6])))
        #right_hand
        right_trans = torch.from_numpy((data[:, 6:9]))
        right_rot = torch.from_numpy((np.radians(data[:, 9:12])))
        #left_hand
        left_trans = torch.from_numpy((data[:, 12:15]))
        left_rot = torch.from_numpy((np.radians(data[:, 15:18])))
        #root
        root_trans = torch.from_numpy((data[:, 18:21]))
        root_rot = torch.from_numpy((np.radians(data[:, 21:24])))
        # left_foot
        left_trans_foot = torch.from_numpy((data[:, 24:27]))
        left_rot_foot = torch.from_numpy((np.radians(data[:, 27:30])))
        # right_foot
        right_trans_foot = torch.from_numpy((data[:, 30:33]))
        right_rot_foot = torch.from_numpy((np.radians(data[:, 33:36])))

        trans_mat = torch.from_numpy(
            np.concatenate((hmd_trans, right_trans, left_trans, root_trans, right_trans_foot, left_trans_foot),
                           axis=1).reshape(-1, feature_bz, 3)).float()
        rot = torch.from_numpy(
            np.concatenate((hmd_rot, right_rot, left_rot, root_rot, right_rot_foot, left_rot_foot),
                           axis=1).reshape(-1,  feature_bz, 3)).float()

        rot_mat = aa2matrot(rot.reshape(-1, 3)).reshape(seq, -1, 3, 3)


        rot_matrot = rot_mat
        trans = trans_mat


        # rot_matrot = aa2matrot(rot.reshape(-1,3)).reshape(seq,-1,9)
        rotation_global_6d = utils_transform.matrot2sixd(rot_matrot.reshape(-1, 3, 3)).reshape(rot_matrot.shape[0], -1, 6)
        rotation_velocity_global_matrot = torch.matmul(torch.inverse(rot_matrot.reshape(-1, feature_bz, 3, 3)[:-1]), rot_matrot.reshape(-1, feature_bz, 3, 3)[1:])
        rotation_velocity_global_6d = utils_transform.matrot2sixd(
            rotation_velocity_global_matrot.reshape(-1, 3, 3)).reshape(rotation_velocity_global_matrot.shape[0], -1, 6)
        data_trans = torch.zeros(seq - 1, 22, 3)
        data_vel = torch.zeros(seq -1, 22, 3)
        data_rot = torch.zeros(seq -1, 22, 6)
        data_rvel = torch.zeros(seq -1, 22, 6)

        indices = np.array([15, 21, 20, 0, 8, 7])
        target = np.array([0, 1, 2, 3, 4, 5])
        data_trans[:, indices, :] = trans[1:, target, :]
        data_vel[:, indices, :] = trans[1:, target, :] - trans[:-1, target, :]
        data_rot[:, indices, :] = rotation_global_6d[1:, target, :]
        data_rvel[:, indices, :] = rotation_velocity_global_6d[:, target]

        hmd_position_global_full_gt_list = torch.cat([data_rot.reshape(seq - 1, -1),
                                                     data_rvel.reshape(seq - 1, -1),
                                                     data_trans.reshape(seq - 1, -1),
                                                     data_vel.reshape(seq - 1, -1)], dim=-1)

        position_head_world = trans[:, 0, ...] # world position of head
        head_global_trans = torch.eye(4).repeat(position_head_world.shape[0], 1, 1)
        head_global_trans[:, :3, :3] = rot_matrot[:, 0, ...].reshape(-1, 3, 3).squeeze()
        head_global_trans[:, :3, 3] = position_head_world
        head_global_trans_list = head_global_trans[1:]


        data_all['hmd_position_global_full_gt_list'] = hmd_position_global_full_gt_list.cpu()
        data_all['head_global_trans_list'] = head_global_trans_list
        data_all['rotation_local_full_gt_list'] = trans
        data_all['body_parms_list'] = rot
        data_all['filepath'] = file_path

        os.makedirs(dst, exist_ok=True)
        idx = npz_file.split('.')[0]
        with open(os.path.join(dst, '{}.pkl'.format(idx)), 'wb') as f:
            pickle.dump(data_all, f)`

[1.zip](https://github.com/user-attachments/files/19279217/1.zip)
